telegram/_utils/warnings_transition.py

```diff
@@ -0,0 +1,126 @@
+# along with this program.  If not, see [http://www.gnu.org/licenses/].
+"""This module contains classes used for transition warnings issued by this library.
+
+It was created to prevent circular imports that would be caused by creating the warnings
+inside warnings.py.
+
+.. versionadded:: NEXT.VERSION
+"""
+import functools
+from collections.abc import Callable
+from typing import Any, Tuple
+
+from telegram._utils.warnings import warn
+from telegram.warnings import PTBDeprecationWarning
+
+
+# Narrower type hints will cause linting errors and/or circular imports.
+# We'll use `Any` here and put type hints in the calling code.
+def warn_about_deprecated_arg_return_new_arg(
+    deprecated_arg: Any,
+    new_arg: Any,
+    deprecated_arg_name: str,
+    new_arg_name: str,
+    bot_api_version: str,
+    stacklevel: int = 3,
+) -> Any:
+    """A helper function for the transition in API when argument is renamed.
+
+    Checks the `deprecated_arg` and `new_arg` objects; warns if non-None `deprecated_arg` object
+    was passed. Returns `new_arg` object (either the one originally passed by the user or the one
+    that user passed as `deprecated_arg`).
+
+    Raises `ValueError` if both `deprecated_arg` and `new_arg` objects were passed, and they are
+    different.
+    """
+    if deprecated_arg and new_arg and deprecated_arg != new_arg:
+        raise ValueError(
+            f"You passed different entities as '{deprecated_arg_name}' and '{new_arg_name}'. "
+            f"The parameter '{deprecated_arg_name}' was renamed to '{new_arg_name}' in Bot API "
+            f"{bot_api_version}. We recommend using '{new_arg_name}' instead of "
+            f"'{deprecated_arg_name}'."
+        )
+
+    if deprecated_arg:
+        warn(
+            f"Bot API {bot_api_version} renamed the argument '{deprecated_arg_name}' to "
+            f"'{new_arg_name}'.",
+            PTBDeprecationWarning,
+            stacklevel=stacklevel,
+        )
+        return deprecated_arg
+
+    return new_arg
+
+
+def warn_about_deprecated_attr_in_property(
+    deprecated_attr_name: str,
+    new_attr_name: str,
+    bot_api_version: str,
+    stacklevel: int = 3,
+) -> None:
+    """A helper function for the transition in API when attribute is renamed.
+    It is meant to be called from properties that replace deprecated attributes in classes.
+    """
+    warn(
+        f"Bot API {bot_api_version} renamed the attribute '{deprecated_attr_name}' to "
+        f"'{new_attr_name}'.",
+        PTBDeprecationWarning,
+        stacklevel=stacklevel,
+    )
+
+
+warn_about_thumb_return_thumbnail = functools.partial(
+    warn_about_deprecated_arg_return_new_arg,
+    deprecated_arg_name="thumb",
+    new_arg_name="thumbnail",
+    bot_api_version="6.6",
+)
+"""A helper function to warn about using a deprecated 'thumb' argument and return it or the new
+'thumbnail' argument, introduced in API 6.6.
+"""
+
+
+def warn_about_required_renamed_param_passed_as_kwarg(
```

Comment:
This function seems somewhat redundant to me 🤔 If you use it, you still need to raise a `ValueError` inside the function if both the deprecated and new (kw)args are passed, so one could as well just use `warn_about_deprecated_arg_return_new_arg` inside of the function. Or am I missing something?

Reply:
hmm... I remember I introduced the decorator after [this discussion](https://t.me/c/1494805131/34757). The tests actually pass without it, but I remember there was an idea behind it. Just have to understand again what my idea was 😆

Reply:
OK I removed the decorator function and all my tests still pass. Which doesn't mean that I didn't miss anything :)

Reply:
You still manually raise `ValueError` at the placeses where the decorator used to be even though that's already covered by `warn_about_deprecated_arg_return_new_arg`

Reply:
My point is that these raisings of Value error can be removed;)

Reply:
Hm... in e.g. `InlineQueryResultGif` I raise `ValueError` when both `thumb` and `thumbnail` are `None`. In `warn_about_deprecated_arg_return_new_arg` I raise `ValueError` when they are both not `None` and are different.